frontend/openchat-agent/src/services/openchatAgent.ts

```diff
@@ -3799,74 +3800,138 @@ export class OpenChatAgent extends EventTarget {
 
             setActivityFeedEvents(sorted.slice(0, MAX_ACTIVITY_EVENTS));
 
-            resolve({ total: server.total, events: sorted }, false);
-
-            const rehydrated = await this.hydrateActivityFeedEvents(sorted).catch((err) => {
-                console.error("Error rehydrating activity feed: ", err);
-                reject(err);
-                return sorted;
-            });
-
-            resolve(
-                {
-                    total: server.total,
-                    events: rehydrated,
-                },
-                true,
+            this.hydrateActivityFeedEvents(sorted, (hydrated, final) =>
+                resolve({ total: server.total, events: hydrated }, final),
             );
         });
     }
 
+    /**
+     * Hydration is a two phase process. First we load all the messages that we have in the cache
+     * and resolve. Secondly we then optionally load any missing messages from the backend and
+     * resolve again.
+     */
     async hydrateActivityFeedEvents(
         activityEvents: MessageActivityEvent[],
-    ): Promise<MessageActivityEvent[]> {
-        // partition the events by message context
+        callback: (events: MessageActivityEvent[], final: boolean) => void,
+    ) {
         const eventIndexesByMessageContext = activityEvents.reduce((map, event) => {
             const eventIndexes = map.get(event.messageContext) ?? [];
             eventIndexes.push(event.eventIndex);
             map.set(event.messageContext, eventIndexes);
             return map;
         }, new AsyncMessageContextMap<number>());
 
-        // look up all the messages correlated with the activity events (mostly this will come from the cache)
+        const [withCachedMessages, missing] = await this.getMessagesByMessageContext(
+            eventIndexesByMessageContext,
+            activityEvents,
+            "cached",
+        );
+
+        const numberOfMissing = [...missing.values()].reduce((total, s) => total + s.size, 0);
```

Comment:
We don't need to calculate `numberOfMissing`, we just need to know if there are any.

Reply:
Maybe something like `const anyMissing = [...missing.values()].some((s) => s.size > 0);`?

Reply:
Yes that's a bit nicer